README.md
# Rust Portable SMBus Packet Error Code Algorithm

[![crates.io](https://img.shields.io/crates/v/smbus-pec.svg)](https://crates.io/crates/smbus-pec)
[![Docs](https://docs.rs/smbus-pec/badge.svg)](https://docs.rs/smbus-pec)
[![Build Status](https://github.com/eldruin/smbus-pec-rs/workflows/Build/badge.svg)](https://github.com/eldruin/smbus-pec-rs/actions?query=workflow%3ABuild)
[![Coverage Status](https://coveralls.io/repos/github/eldruin/smbus-pec-rs/badge.svg?branch=master)](https://coveralls.io/github/eldruin/smbus-pec-rs?branch=master)

This is a portable minimal implementation of the [System Management Bus (SMBus)][smbus]
[Packet Error Code][smbus-pec] calculation algorithm intended for use in `no_std`.

SMBus 1.1 and later defines an optional Packet Error Checking mode. When used, an extra byte
is appended to all transmissions containing a Packet Error Code (PEC).

The PEC is calculated over the whole transmission including address and read/write bit.

The polynomial used is `x^8 + x^2 + x + 1`, which corresponds to [CRC-8-ATM HEC][crc8]
initialized to zero.

## How this crate compares to others

There is a number of crates implementing CRC algorithms but their intention is to
be configurable, generic, use acceleration via SIMD instructions, etc.

This crate provides a portable and non-configurable implementation of exactly one
algorithm: The one used for SMBus PEC (optionally using a pre-calculated lookup table).

This should allow the compiler to make good optimizations and allows for use of the
algorithm in any target architecture with minimal code bloat.

This makes this crate specially well suited for use in `no_std` environments.

## Pre-calculated lookup table

A faster version of the algorithm is provided through the use of a pre-calculated
lookup table. This can be enabled through the `lookup-table` feature.

With this feature enabled a table of 256 pre-calculated `u8` values will be included
which avoids bit-by-bit calculation at the cost of the space needed to store it.

## Usage

```rust
use smbus_pec::pec;

const ADDRESS: u8 = 0x5A;
const REGISTER: u8 = 0x06;

fn main() {
    let pec_write = pec(&[ADDRESS << 1, REGISTER, 0xAB, 0xCD]);
    println!("PEC: {}", pec_write); // prints 95

    let data = [ADDRESS << 1, REGISTER, (ADDRESS << 1) + 1, 38, 58];
    let pec_write_read = pec(&data);
    println!("PEC: {}", pec_write_read); // prints 102
}
```

## Support

For questions, issues, feature requests, other changes, or just feedback, please file an
[issue in the github project](https://github.com/eldruin/smbus-pec-rs/issues).

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)

at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.

[crc8]: https://en.wikipedia.org/wiki/CRC-8
[smbus]: https://en.wikipedia.org/wiki/System_Management_Bus
[smbus-pec]: https://en.wikipedia.org/wiki/System_Management_Bus#Packet_Error_Checking


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

...

## [1.0.1] - 2021-05-23

### Fixed
- Avoid unnecessary recompilations.

## [1.0.0] - 2021-05-13

### Changed
- Updated `embedded-crc-macros` dependency to version `1.0.0`.

## [0.1.0] - 2020-08-09

Initial release to crates.io.

[Unreleased]: https://github.com/eldruin/smbus-pec-rs/compare/v1.0.1...HEAD
[1.0.1]: https://github.com/eldruin/smbus-pec-rs/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/eldruin/smbus-pec-rs/compare/v0.1.0...v1.0.0
[0.1.0]: https://github.com/eldruin/smbus-pec-rs/releases/tag/v0.1.0

tests/integration.rs
use core::hash::Hasher;
use smbus_pec::{pec, Pec};
const ADDR: u8 = 0x5A;
const REGISTER: u8 = 0x06;

const TABLE: [u8; 256] = [
    0x00, 0x07, 0x0E, 0x09, 0x1C, 0x1B, 0x12, 0x15, 0x38, 0x3F, 0x36, 0x31, 0x24, 0x23, 0x2A, 0x2D,
    0x70, 0x77, 0x7E, 0x79, 0x6C, 0x6B, 0x62, 0x65, 0x48, 0x4F, 0x46, 0x41, 0x54, 0x53, 0x5A, 0x5D,
    0xE0, 0xE7, 0xEE, 0xE9, 0xFC, 0xFB, 0xF2, 0xF5, 0xD8, 0xDF, 0xD6, 0xD1, 0xC4, 0xC3, 0xCA, 0xCD,
    0x90, 0x97, 0x9E, 0x99, 0x8C, 0x8B, 0x82, 0x85, 0xA8, 0xAF, 0xA6, 0xA1, 0xB4, 0xB3, 0xBA, 0xBD,
    0xC7, 0xC0, 0xC9, 0xCE, 0xDB, 0xDC, 0xD5, 0xD2, 0xFF, 0xF8, 0xF1, 0xF6, 0xE3, 0xE4, 0xED, 0xEA,
    0xB7, 0xB0, 0xB9, 0xBE, 0xAB, 0xAC, 0xA5, 0xA2, 0x8F, 0x88, 0x81, 0x86, 0x93, 0x94, 0x9D, 0x9A,
    0x27, 0x20, 0x29, 0x2E, 0x3B, 0x3C, 0x35, 0x32, 0x1F, 0x18, 0x11, 0x16, 0x03, 0x04, 0x0D, 0x0A,
    0x57, 0x50, 0x59, 0x5E, 0x4B, 0x4C, 0x45, 0x42, 0x6F, 0x68, 0x61, 0x66, 0x73, 0x74, 0x7D, 0x7A,
    0x89, 0x8E, 0x87, 0x80, 0x95, 0x92, 0x9B, 0x9C, 0xB1, 0xB6, 0xBF, 0xB8, 0xAD, 0xAA, 0xA3, 0xA4,
    0xF9, 0xFE, 0xF7, 0xF0, 0xE5, 0xE2, 0xEB, 0xEC, 0xC1, 0xC6, 0xCF, 0xC8, 0xDD, 0xDA, 0xD3, 0xD4,
    0x69, 0x6E, 0x67, 0x60, 0x75, 0x72, 0x7B, 0x7C, 0x51, 0x56, 0x5F, 0x58, 0x4D, 0x4A, 0x43, 0x44,
    0x19, 0x1E, 0x17, 0x10, 0x05, 0x02, 0x0B, 0x0C, 0x21, 0x26, 0x2F, 0x28, 0x3D, 0x3A, 0x33, 0x34,
    0x4E, 0x49, 0x40, 0x47, 0x52, 0x55, 0x5C, 0x5B, 0x76, 0x71, 0x78, 0x7F, 0x6A, 0x6D, 0x64, 0x63,
    0x3E, 0x39, 0x30, 0x37, 0x22, 0x25, 0x2C, 0x2B, 0x06, 0x01, 0x08, 0x0F, 0x1A, 0x1D, 0x14, 0x13,
    0xAE, 0xA9, 0xA0, 0xA7, 0xB2, 0xB5, 0xBC, 0xBB, 0x96, 0x91, 0x98, 0x9F, 0x8A, 0x8D, 0x84, 0x83,
    0xDE, 0xD9, 0xD0, 0xD7, 0xC2, 0xC5, 0xCC, 0xCB, 0xE6, 0xE1, 0xE8, 0xEF, 0xFA, 0xFD, 0xF4, 0xF3,
];

#[test]
fn check_pec_table() {
    for (i, expected) in TABLE.iter().enumerate() {
        assert_eq!(pec(&[i as u8]), *expected);
    }
}

#[test]
fn check_pec_array() {
    assert_eq!(pec(&[ADDR << 1, REGISTER, 0xAB, 0xCD]), 95);
    assert_eq!(pec(&[ADDR << 1, REGISTER, (ADDR << 1) + 1, 38, 58]), 102);
    assert_eq!(pec(&[ADDR << 1, REGISTER, (ADDR << 1) + 1, 107, 58]), 212);
    assert_eq!(pec(&[ADDR << 1, REGISTER, (ADDR << 1) + 1, 97, 58]), 86);
    assert_eq!(pec(&[ADDR << 1, REGISTER, (ADDR << 1) + 1, 225, 57]), 233);
}

#[test]
fn check_pec_hasher_table() {
    for (i, expected) in TABLE.iter().enumerate() {
        let mut hasher = Pec::new();
        hasher.write(&[i as u8]);
        assert_eq!(hasher.finish(), *expected as u64);
    }
}

#[test]
fn check_pec_hasher_array() {
    let mut hasher = Pec::new();
    hasher.write(&[ADDR << 1, REGISTER, 0xAB, 0xCD]);
    assert_eq!(hasher.finish(), 95);
}


benches/benchmarks.rs
use core::hash::Hasher;
use criterion::{black_box, criterion_group, criterion_main, BenchmarkId, Criterion, Throughput};
use rand::{rngs::SmallRng, FromEntropy, Rng};
use smbus_pec::{pec, Pec};

fn fill_random(v: &mut Vec<u8>) {
    let mut rng = SmallRng::from_entropy();
    for _ in 0..v.capacity() {
        v.push(rng.gen::<u8>());
    }
}

const MESSAGE_SIZES: [usize; 7] = [1, 3, 5, 20, 50, 100, 500];
fn bench(c: &mut Criterion) {
    let mut group = c.benchmark_group("pec");
    for size in &MESSAGE_SIZES {
        group.throughput(Throughput::Bytes(*size as u64));
        let mut data: Vec<u8> = Vec::with_capacity(*size);
        fill_random(&mut data);
        group.bench_with_input(BenchmarkId::new("fn", *size), &data, |b, data| {
            b.iter(|| {
                let pec = pec(data);
                black_box(&pec);
            });
        });
        group.bench_with_input(BenchmarkId::new("hasher", *size), &data, |b, data| {
            b.iter(|| {
                let mut hasher = Pec::new();
                hasher.write(data);
                let pec = hasher.finish();
                black_box(&pec);
            });
        });
    }
    group.finish();
}

criterion_group!(benches, bench);
criterion_main!(benches);


src/lib.rs
//! This is a portable minimal implementation of the [System Management Bus (SMBus)][smbus]
//! [Packet Error Code][smbus-pec] calculation algorithm intended for use in `no_std`.
//!
//! SMBus 1.1 and later defines an optional Packet Error Checking mode. When used, an extra byte
//! is appended to all transmissions containing a Packet Error Code (PEC).
//!
//! The PEC is calculated over the whole transmission including address and read/write bit.
//! The polynomial used is `x^8 + x^2 + x + 1`, which corresponds to [CRC-8-ATM HEC][crc8]
//! initialized to zero.
//!
//! [crc8]: https://en.wikipedia.org/wiki/CRC-8
//! [smbus]: https://en.wikipedia.org/wiki/System_Management_Bus
//! [smbus-pec]: https://en.wikipedia.org/wiki/System_Management_Bus#Packet_Error_Checking
//!
//! # How this crate compares to others
//!
//! There is a number of crates implementing CRC algorithms but their intention is to
//! be configurable, generic, use acceleration via SIMD instructions, etc.
//!
//! This crate provides a portable and non-configurable implementation of exactly one
//! algorithm: The one used for SMBus PEC (optionally using a pre-calculated lookup table).
//!
//! This should allow the compiler to make good optimizations and allows for use of the
//! algorithm in any target architecture with minimal code bloat.
//!
//! This makes this crate specially well suited for use in `no_std` environments.
//!
//! # Pre-calculated lookup table
//!
//! A faster version of the algorithm is provided through the use of a pre-calculated
//! lookup table. This can be enabled through the `lookup-table` feature.
//!
//! With this feature enabled a table of 256 pre-calculated `u8` values will be included
//! which avoids bit-by-bit calculation at the cost of the space needed to store it.
//!
//! # Usage
//!
//! ## Send command with a value to an address using function
//!
//! A typical use case for this would be writing the value to a register
//!
//! ```rust
//! # struct Smbus;
//! # impl Smbus {
//! #   fn write(&mut self, address: u8, data: &[u8]) {}
//! # }
//! #
//! # let mut smbus = Smbus {};
//! #
//! use smbus_pec::pec;
//! const ADDR: u8 = 0x5A;
//! let command = 0x06;
//! let value = 0xAB;
//!
//! let checksum = pec(&[ADDR << 1, command, value]);
//! smbus.write(ADDR, &[command, value, checksum]);
//! ```
//!
//! ## Perform request-response using function
//!
//! A typical use case for this would be reading the value of a register.
//! To do this, a write with the register address is sent, followed by a read.
//!
//! ```rust
//! # struct Smbus;
//! # impl Smbus {
//! #   fn write_read(&mut self, address: u8, data: &[u8], buffer: &mut [u8]) {}
//! # }
//! #
//! # let mut smbus = Smbus {};
//! #
//! use smbus_pec::pec;
//! const ADDR: u8 = 0x5A;
//! let register = 0x06;
//! let mut data = [0; 2];
//!
//! smbus.write_read(ADDR, &[register], &mut data);
//! let checksum = pec(&[ADDR << 1, register, (ADDR << 1) | 1, data[0]]);
//! if checksum != data[1] {
//!   println!("Packet Error Code mismatch.");
//! }
//! let value = data[0];
//! ```
//!
//! ## Send command with a value to an address using hasher
//!
//! A typical use case for this would be writing the value to a register
//!
//! ```rust
//! # struct Smbus;
//! # impl Smbus {
//! #   fn write(&mut self, address: u8, data: &[u8]) {}
//! # }
//! #
//! # let mut smbus = Smbus {};
//! #
//! use core::hash::Hasher;
//! use smbus_pec::Pec;
//! const ADDR: u8 = 0x5A;
//! let command = 0x06;
//! let value = 0xAB;
//!
//! let mut hasher = Pec::new();
//! hasher.write(&[ADDR << 1, command, value]);
//! let checksum = hasher.finish() as u8;
//! smbus.write(ADDR, &[command, value, checksum]);
//! ```
//!
//! ## Perform request-response using hasher
//!
//! A typical use case for this would be reading the value of a register.
//! To do this, a write with the register address is sent, followed by a read.
//!
//! ```rust
//! # struct Smbus;
//! # impl Smbus {
//! #   fn write_read(&mut self, address: u8, data: &[u8], buffer: &mut [u8]) {}
//! # }
//! #
//! # let mut smbus = Smbus {};
//! #
//! use core::hash::Hasher;
//! use smbus_pec::Pec;
//! const ADDR: u8 = 0x5A;
//! let register = 0x06;
//! let mut data = [0; 2];
//!
//! smbus.write_read(ADDR, &[register], &mut data);
//! let mut hasher = Pec::new();
//! hasher.write(&[ADDR << 1, register, (ADDR << 1) | 1, data[0]]);
//! let checksum = hasher.finish();
//! if checksum != data[1] as u64 {
//!   println!("Packet Error Code mismatch.");
//! }
//! let value = data[0];
//! ```
//!

#![doc(html_root_url = "https://docs.rs/smbus-pec/1.0.1")]
#![deny(unsafe_code, missing_docs)]
#![no_std]

#[cfg(not(feature = "lookup-table"))]
mod default_impl {
    use embedded_crc_macros::{crc8, crc8_hasher};
    crc8!(
        fn pec,
        7,
        0,
        "Calculate SMBus Packet Error Code over transmitted data.\
        \n\nThe input data array must contain the complete message including address and read/write bit."
    );

    crc8_hasher!(
        struct Pec,
        7,
        0,
        "Calculate SMBus Packet Error Code over transmitted data. `core::hash::Hasher` implementation.\
        \n\nThe input data array must contain the complete message including address and read/write bit.\
        \n\nNote that the hasher holds an internal state so that it is possible to call `write()` \
        several times. A new instance must be created for each independent calculation."
    );
}

#[cfg(not(feature = "lookup-table"))]
pub use crate::default_impl::{pec, Pec};

#[cfg(feature = "lookup-table")]
mod lookup_table_impl {
    use crate::LOOKUP_TABLE;
    use embedded_crc_macros::{crc8_hasher_lookup_table, crc8_lookup_table};

    crc8_lookup_table!(
        fn pec,
        0,
        LOOKUP_TABLE,
        "Calculate SMBus Packet Error Code over transmitted data.\
        \n\nThe input data array must contain the complete message including address and read/write bit.
        \n\nThis implementation uses a lookup table and is much faster at the cost of some space."
    );

    crc8_hasher_lookup_table!(
        struct Pec,
        0,
        LOOKUP_TABLE,
        "Calculate SMBus Packet Error Code over transmitted data. `core::hash::Hasher` implementation.\
        \n\nThe input data array must contain the complete message including address and read/write bit.\
        \n\nNote that the hasher holds an internal state so that it is possible to call `write()` \
        several times. A new instance must be created for each independent calculation."
    );
}

#[cfg(feature = "lookup-table")]
pub use crate::lookup_table_impl::{pec, Pec};

#[cfg(feature = "lookup-table")]
include!(concat!(env!("OUT_DIR"), "/lookup_table.rs"));
